bigchaindb_driver/utils.py

```diff
@@ -60,20 +59,44 @@ def _get_default_port(scheme):
     return 443 if scheme == 'https' else 9984
 
 
-def _normalize_nodes(*nodes):
-    if not nodes:
-        return (DEFAULT_NODE,)
+def _normalize_url(node):
+    parts = urlparse(node, scheme='http', allow_fragments=False)
+    port = parts.port if parts.port else _get_default_port(parts.scheme)
+    netloc = '{}:{}'.format(parts.hostname, port)
+    return urlunparse((parts.scheme, netloc, parts.path, '', '', ''))
+
+
+def _normalize_dict(*nodes):
+    norm_nodes = []
+    for node in nodes:
+        if not node:
+            norm_nodes.append(
+                {"endpoint": DEFAULT_NODE, "headers": node["headers"]})
+            continue
+        elif '://' not in node["endpoint"]:
+            node["endpoint"] = '//{}'.format(node["endpoint"])
+        url = _normalize_url(node["endpoint"])
+        norm_nodes.append({"endpoint": url, "headers": node["headers"]})
+    return tuple(norm_nodes,)
+
+
+def _normalize_array(*nodes):
     norm_nodes = []
     for node in nodes:
         if not node:
             norm_nodes.append(DEFAULT_NODE)
             continue
         elif '://' not in node:
             node = '//{}'.format(node)
-        parts = urlparse(node, scheme='http', allow_fragments=False)
-        port = parts.port if parts.port else _get_default_port(parts.scheme)
-        netloc = '{}:{}'.format(parts.hostname, port)
-        norm_nodes.append(
-            urlunparse((parts.scheme, netloc, parts.path, '', '', '')))
-
+        url = _normalize_url(node)
+        norm_nodes.append(url)
     return tuple(norm_nodes,)
+
+
+def _normalize_nodes(*nodes):
+    if not nodes:
+        return (DEFAULT_NODE,)
+    if isinstance(nodes[0], dict):
```

Comment:
The best practice is to use `issubclass`, so in case someone is passing a custom object that inherits from `dict`, this check will still work. (I doubt this will ever create problems, but _safe is better than sorry_.)